### PR TITLE
Allow users to control the maximum number of nesting levels generated for the pup functions

### PR DESCRIPTION
**Pull request by pavanbalaji**
## Pull Request Description

Allow for users to generate different levels of nesting, thus trading off performance with the library size and build time.  The default is set to 3 levels of nesting, as before.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [x] Have read and agree to the Yaksa CLA terms (https://github.com/pmodels/yaksa/wiki/Yaksa-Contributor-License-Agreement)

### CODE REVIEW

File: autogen.sh [hunk 1]
```diff
@@ -14,6 +18,24 @@ error() {
...
+## Parse user arguments
+########################################################################
+
+pup_max_nesting=3
+while test ! -z "$1" ; do
+    if test "$1" = "--pup-max-nesting" ; then
+        shift
+        pup_max_nesting=$1
+	shift
+    fi
```
Comment (hzhou) [inline autogen.sh:31]: It's a dead loop if wrong argument is supplied since `$1` will never get shifted.

File: autogen.sh [hunk 2]
```diff
@@ -22,10 +44,14 @@ echo "=== done === "
...
 
+########################################################################
+## Generating required files
+########################################################################
+
 # backend pup functions
 for x in seq cuda ; do
     echo_n "generating backend pup functions for ${x}... "
-    ./src/backend/${x}/genpup.py
+    ./src/backend/${x}/genpup.py --pup-max-nesting=${pup_max_nesting}
```
Comment (hzhou) [inline autogen.sh:54]: Would it be simpler if you use the same option syntax for `autogen.sh` and `genpup.py`? You can use a case switch and collect the arguments as is and pass them to `genpup.py`.

File: src/backend/cuda/genpup.py [hunk 3]
```diff
@@ -320,7 +321,16 @@ def switcher(typelist, pupstr, nests):
 ##### main function
 ########################################################################################
 if __name__ == '__main__':
-    #### generate the core pack/unpack kernels
+    ##### parse user arguments
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--pup-max-nesting', type=int, help='maximum nesting levels to generate', required=True)
+    args = parser.parse_args()
```
Comment (hzhou) [inline src/backend/cuda/genpup.py:327]: How about add some sanity check so the nesting level is not negative or too big? I think it is nice to have at least a warning when the nesting level is above e.g. 4 so make sure the user is aware of what is going to happen.

File: autogen.sh (hunk 1 above)
Comment (pavanbalaji) [reply autogen.sh:31]: Good catch.  I'll fix it.

File: autogen.sh (hunk 2 above)
Comment (pavanbalaji) [reply autogen.sh:54]: For now, the arguments are only for `genpup.py`, but `autogen.sh` runs more than that one script (for example, `gentests.py`).  So we can't simply pass all arguments to `genpup.py`.

File: src/backend/cuda/genpup.py (hunk 3 above)
Comment (pavanbalaji) [reply src/backend/cuda/genpup.py:327]: I can check for negative values.  But "too large" is an arbitrary number, which I don't want to add.

File: autogen.sh (hunk 2 above)
Comment (hzhou) [reply autogen.sh:57]: I understand. I mean something like this:
```
genpup_args=
for a in $@ ; do
    case a in 
        --pup-*) genpup_args="$genpup_args $a" ;;
    esac
done
```
then
```
    eval ./src/.../genpup.py $genpup_args
```

Comment (pavanbalaji) [reply autogen.sh:57]: Yeah, that's a good idea.  It'll also naturally separate the namespace for the arguments for different scripts.  I'll fix it.

File: src/backend/seq/genpup.py [hunk 4]
```diff
@@ -340,7 +341,21 @@ def switcher(typelist, pupstr, nests):
...
+        parser.print_help()
+        print
+        print("===> ERROR: pup-max-nesting must be positive")
+        sys.exit(1)
+
+    ##### generate the list of derived datatype arrays
+    darraylist = [ ]
+    yutils.generate_darrays(derived_types, darraylist, args.pup_max_nesting)
+
+    ##### generate the core pack/unpack kernels
```
Comment (hzhou) [inline src/backend/seq/genpup.py:358]: Is it easy to factor out a common module and avoid duplication?

Comment (pavanbalaji) [reply src/backend/seq/genpup.py:358]: `yutils` is the common module.  It's meant for functions that are generic across all `yaksa` scripts.

**Pull request merged**